### PR TITLE
feat: add mfa session to secure otp login

### PR DESCRIPTION
**Pull request by catusax**
#### What does this PR do?

this pr adds a mfa session cookie when mfa is enabled at the login step.

in the previous version , server uses otp table to make sure a user has passed password check. but it did not check if the client which passed password check and the client which has otp code are same client. so, when a client has logged in with password and not input otp code yet, another client which did not know password can call VerifyOtp and collide the otp code to login.

with mfa session, we can also support totp or webauth.  mfa session can be used to check if password login passed before totp and webauth

### CODE REVIEW

**Comment by lakhansamani** [comment]
@catusax I like the idea, but this might not fly for mobile version.
There are people trying to use it with react native.

Any thoughts for that?

File: server/constants/cookie.go [hunk 1]
```diff
@@ -5,4 +5,6 @@ const (
 	AppCookieName = "cookie"
 	// AdminCookieName is the name of the cookie that is used to store the admin token
 	AdminCookieName = "authorizer-admin"
+
+	MfaCookieName = "mfa"
```
Comment (lakhansamani) [inline server/constants/cookie.go:9]: Pls add comment for exported variables

File: server/cookie/mfa_session.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/authorizerdev/authorizer/server/constants"
+	"github.com/authorizerdev/authorizer/server/memorystore"
+	"github.com/authorizerdev/authorizer/server/parsers"
+	"github.com/gin-gonic/gin"
+)
+
+// SetSession sets the session cookie in the response
```
Comment (lakhansamani) [inline server/cookie/mfa_session.go:15]: This should be // SetMfaSession...

File: server/cookie/mfa_session.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+		gc.SetSameSite(http.SameSiteNoneMode)
+	}
+	// TODO allow configuring from dashboard
+	age := 60
+
+	gc.SetCookie(constants.MfaCookieName+"_session", sessionID, age, "/", host, secure, httpOnly)
+	gc.SetCookie(constants.MfaCookieName+"_session_domain", sessionID, age, "/", domain, secure, httpOnly)
+}
+
+// DeleteSession sets session cookies to expire
```
Comment (lakhansamani) [inline server/cookie/mfa_session.go:50]: This should be // DeleteMfaSession ...

File: server/memorystore/providers/inmemory/store.go [hunk 4]
```diff
@@ -42,6 +42,24 @@ func (c *provider) DeleteSessionForNamespace(namespace string) error {
 	return nil
 }
 
+func (c *provider) SetMfaSession(email, key string, expiration int64) error {
```
Comment (lakhansamani) [inline server/memorystore/providers/inmemory/store.go:45]: Pls add comments for all 3

File: server/memorystore/providers/providers.go [hunk 5]
```diff
@@ -13,6 +13,10 @@ type Provider interface {
 	// DeleteSessionForNamespace deletes the session for a given namespace
 	DeleteSessionForNamespace(namespace string) error
 
+	SetMfaSession(email, key string, expiration int64) error
```
Comment (lakhansamani) [inline server/memorystore/providers/providers.go:16]: Pls add comments for all 3

File: server/memorystore/providers/redis/store.go [hunk 6]
```diff
@@ -91,6 +93,34 @@ func (c *provider) DeleteSessionForNamespace(namespace string) error {
 	return nil
 }
 
+func (c *provider) SetMfaSession(email, key string, expiration int64) error {
```
Comment (lakhansamani) [inline server/memorystore/providers/redis/store.go:96]: Pls add comment for all 3

**Review by lakhansamani — COMMENTED** [review]
Some minor changes with commenting

**Comment by catusax** [comment]
seems VerifyOtpResolver supports phone_number now, i will resolve the conflict when i have time.

about react native, im not a frontend developer, but this pr adds a new cookie, if the react native uses cookie, it should be working just fine. if not, then we need to add a new field to `AuthResponse`

**Comment by lakhansamani** [comment]
@catusax is this PR ready for reviewing again?

**Comment by catusax** [comment]
@lakhansamani yes

**Review by lakhansamani — APPROVED** [review]

**Comment by lakhansamani** [comment]
Thanks @catusax for the contribution this will be released as part of 1.1.61

**Pull request merged**